### PR TITLE
fix(no-effect-chain): preserve stable callback precision

### PR DESCRIPTION
**Pull request by aidenybai**
## Summary

- follow synchronously invoked callbacks proven to originate from React `useCallback` for reader reachability, cleanup detection, and external-system synchronization
- infer callback-hidden state writes only when exact setter symbols prove one state target and the reachable callback graph contains no opaque call or other observable work
- propagate proven helper arguments, guards, immutable aliases, React dependency containers, browser/global receivers, and React-ref provenance
- add paired regression tests and permanent fuzz seeds for every widened precision boundary

## Grounding

The original false positive is the pinned React Bench SlideshowContext job `fix-react-rdh-igordanchenko-yet__4KfhrDj`. A finite playback transition updates `playing`, while the downstream effect synchronizes a timer and React ref. The direct state write was reported even though the accepted callback-extracted implementation performs the same bounded transition.

An earlier version of this PR followed every stable callback writer and added real false positives in AppFlowy, Medusa, and Payload. This version removes that syntax-based widening:

- AppFlowy contains opaque Yjs/context-backed row loading.
- Medusa atomically co-writes two correlated range states.
- Payload performs Lexical DOM synchronization before selecting an item.

All three stay quiet without library names or API-spelling exceptions. A direct, aliased, optional, nested, or conditional stable callback that provably performs only one React state transition still reports.

## Precision boundary

- React `useCallback` origin and state setters require exact binding-symbol identity.
- Only synchronous local call edges are traversed; async functions, generators, deferred callbacks, mutable aliases, and unresolved imports remain conservative.
- Stable writer inference requires exactly one distinct state target across the full reachable graph.
- Opaque calls, assignments, updates, property mutation, constructors, throws, tagged templates, deletes, dynamic imports, and executable default parameters reject stable writer proof.
- External synchronization requires proven browser/global/import/React-ref/dependency provenance; similarly named userland APIs and shadowed globals do not qualify.
- Cleanup results are classified through reachable return paths, including concise, optional, conditional, and helper-returned forms.

## Validation

Current exact head: `fb0e04ff2d5f9b29af936a72ce270913fa1489b9`. The detector validation below ran on `09a9bf1216198e5bc523e7a74cc3a69d79f093fc`; the only later change removes non-portable wall-clock assertions from two unchanged deep traversal fixtures after five CI runners exceeded the absolute threshold while passing their semantic assertions.

- full repository tests: 14/14 tasks
- plugin suite: 573 files, 15,328 passed, 188 skipped
- typecheck: 15/15 tasks
- lint, format check (3,224 files), JSON-report smoke test, and `git diff --check`
- post-change `truffler` audit: no duplicate helper
- target corpus: 29 true-positive files produce 30 findings; 26 regression files produce zero; no parse failures
- `FUZZ_RULE=no-effect-chain FUZZ_STRICT=1 FUZZ_ITERATIONS=500 nr fuzz`
- exact pinned AppFlowy, both Medusa files, and Payload: zero target diagnostics; paired one-state controls report

### RDE

- 101/101 pinned roots, 13 repositories
- 25,771 files per version
- main / prior PR head / candidate: 18 / 18 / 18 findings
- zero parse or scan failures; zero added or removed verdicts

### React Bench

- 125/125 pinned repositories, 120,711/120,711 files, zero scan failures
- main / prior PR head / candidate: 188 / 195 / 174 findings
- candidate versus main: 0 added, 14 removed
- candidate versus prior: 0 added, 21 removed
- every removed verdict manually classified as a false positive

### Benchmark bases and oracles

- 111/111 bases and 112/112 solutions
- 111,153 base files and 112,904 solution files; zero task, patch, or scan failures
- base counts: 167 / 174 / 155
- solution counts: 164 / 171 / 152
- candidate versus main: 0 added, 24 removed
- candidate versus prior: 0 added, 38 removed
- every removed verdict manually classified as a false positive

The PR remains draft while exact-head CI and automated review settle.

---

> [!NOTE]
> **Medium Risk**
> Changes static-analysis soundness for a core React effects rule; mis-tuned reachability could suppress real chains or reintroduce noise, though the large paired corpus targets that boundary.
> 
> **Overview**
> This patch release tightens **`no-effect-chain`** so legitimate effect chains that sync through stable **`useCallback`** helpers (timers, storage, HTTP, query clients, refs, cleanup) are less often flagged, while still reporting true redundant state-to-state chains.
> 
> The analyzer is described as following **synchronous** `useCallback` bodies for reader reachability and external-sync detection, treating callback-hidden state writes as benign only when the reachable graph targets **one** state setter and has no opaque or observable side work. **Global and import receiver provenance** is narrowed to direct, non-defaulted bindings (including TS wrapper transparency for mutations), with paired fuzz regressions and true-positive controls for alias, control-flow, and name-heuristic edges.
> 
> **`no-chain-state-updates`** now resolves `useCallback` origins via the shared **`getReactUseCallbackCall`** helper instead of inlined alias logic. **`js-min-max-loop`** gains regression coverage for **destructured `globalThis` Math** aliases (immutable direct bindings vs conservative indirect cases). Fuzz **effect snippet pools** are expanded to mirror the new precision boundaries.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit bce0d49c575c76a44f375eaa9fa118a8d244a14c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-effect-chain.ts [hunk 1]
```diff
@@ -524,6 +674,13 @@ const isFunctionShapedReturn = (
       if (setterToStateName.has(returnedValue.callee.name)) return false;
       if (isSetterIdentifier(returnedValue.callee.name)) return true;
     }
+    const invokedFunction = resolveSynchronouslyInvokedFunction(returnedValue.callee, scopes);
+    if (
+      invokedFunction &&
+      returnsOnlyNonCleanupValues(invokedFunction, setterToStateName, scopes)
+    ) {
+      return false;
+    }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-effect-chain.ts:1631]: Confirmed on exact pushed head a7b8c057: `ChainExpression` was not unwrapped before `isFunctionShapedReturn`. Fixed in 06b0d0045 by applying the existing transparent-expression unwrap before return-shape analysis. Added the paired regression `return subscribeAndCopy?.()` (real cleanup stays quiet); the existing optional stable state-copy case still reports. Validation: `nr --filter oxlint-plugin-react-doctor test -- no-effect-chain.regressions.test.ts` (14,874 pass), `nr test` (14/14 tasks), `nr lint`, `nr typecheck`, `nr format:check`, strict generated and React Bench corpus fuzz at 500 iterations, the 40-job Slideshow matrix (0 additions), all 32 React Bench 5 oracles (0/0), and direct AST RDE on 100/100 pinned roots with 0 parse failures and 0 target-rule deltas.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-effect-chain.ts [hunk 2]
```diff
@@ -519,20 +670,30 @@ const isFunctionShapedReturn = (
...
+      scopes,
+    );
+    if (
+      invokedFunction &&
+      returnsOnlyNonCleanupValues(invokedFunction, setterToStateName, scopes)
+    ) {
+      return false;
     }
-    return isCleanupReturn(returnedValue, EMPTY_CLEANUP_NAME_SET, EMPTY_CLEANUP_NAME_SET, {
+    return isCleanupReturn(unwrappedReturnedValue, EMPTY_CLEANUP_NAME_SET, EMPTY_CLEANUP_NAME_SET, {
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-effect-chain.ts:1643]: Confirmed on exact head 06b0d004: the concise effect path used `allowOpaqueReturn: false` without proving the resolved stable callback returns cleanup. Fixed in 6f408ef1a with narrow return-path proof: every explicit return must be a proven cleanup or undefined/fallthrough, and at least one path must return cleanup. Opaque calls and mixed cleanup/object branches still report. Added regressions for concise cleanup, cleanup-or-bare/global-undefined, opaque result, and mixed invalid branches, plus a permanent fuzz seed. Validation: plugin suite 14,879 pass/188 skip; full `nr test` 14/14; lint/typecheck/format/diff green; strict generated and React Bench fuzz 500 each; 40-job Slideshow matrix 0 additions; all 32 Bench 5 oracles 0/0; direct AST RDE 100/100 roots, 0 parse failures, 18→18 target findings, 0 deltas.

File: packages/oxlint-plugin-react-doctor/src/plugin/utils/is-proven-global-namespace-reference.ts [hunk 3]
```diff
@@ -34,26 +62,39 @@ export const isProvenGlobalNamespaceReference = (
...
+      isNodeOfType(declaration.id, "ObjectPattern") &&
+      isNodeOfType(bindingProperty, "Property") &&
+      bindingProperty.value === bindingIdentifier &&
+      bindingProperty.parent === declaration.id &&
+      declaration.init
+    ) {
+      return isProvenGlobalObjectReference(declaration.init, scopes);
+    }
+    if (!symbol.initializer) return false;
+    currentExpression = symbol.initializer;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/utils/is-proven-global-namespace-reference.ts]: Fixed in bce0d49c5. Global namespace provenance now follows only direct const identifier aliases after the explicit safe destructuring case, so a default expression from an unknown object cannot masquerade as the runtime value. Added the reported defaulted-destructuring regression; 453 focused tests, typecheck, strict 1,000-iteration fuzz, lint, and format all pass.
